text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+- The `mod-pubsub` module is also responsible for maintaining a trail of activities: event publication; subscription distribution; errors; non-delivery; etc. By using the correlationID embedded in events it is possible to reconstruct the sequence of operations that form a particular workflow.
+
+- By configuring specific event subscriptions for each tenant it is possible to implement specific cases of inter-app coordination.
+
+- The `mod-pubsub` module is implemented using a message queue (Kafka) which will provide the necessary persistence.
+
+#### High-level solution structure
+There is a global FOLIO interface `pubsub` defined in the scope of this solution which provides APIs to ensure the functioning of the subsystem:
+- API for event type registration by publishers and subscribers. This API must be called by back-end modules.
+  - `POST /pubsub/eventtypes` - to register an event type 
+  - `DELETE /pubsub/eventtypes` - to remove event type registration
+  - `POST /pubsub/eventtypes/declare/publisher` - to declare a publisher for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/publisher` - to remove publisher declaration
+  - `POST /pubsub/eventtypes/declare/subscriber` - to declare a subscriber for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/subscriber` - to remove subscriber declaration
+- API to retrieve existing registered event types.
+  - `GET /pubsub/eventtypes` - returns all registered event types
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}` - returns an event descriptor for a particular event type
+- API to retrieve the registered subscriptions for a specified tenant and event type
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/subscribers` - returns all subscribers of the event type provided in the path variable
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/publishers` - returns all publishers of the event type provided in the path variable
+- API used by publishers to send events. This API must be called by back-end modules which act as publishers
+  - `POST /pubsub/publish`
+- API to retrieve the activity history performed by the event manager. A correlationID can optionally be provided to limit context.
+  - `GET /pubsub/history`
+  - `GET /pubsub/history?corellationId={corellationId}`
+
+It is allowed for publishers to provide callback URLs to receive feedback from `mod-pubsub` module in cases when:
+- Subscriber did not receive an event within the specified period of time.
+- All subscribers deleted their subscriptions on a particular event type registered by this publisher.
+ 
+
+There is a set of requirements that must be followed by a module to provide “Subscriber endpoints” to consume events. The basic set of requirements:
```

Comment:
What should happen if the endpoint specified does not meet these expectations?

When might the system discover this?

Reply:
What are the expectations that `mod-pubsub` has for the response from a delivery endpoint?

Reply:
If delivery fails, will `mod-pubsub` attempt to retry? 

Potentially delivering a message more than once, meaning subscribers might need to be idempotent for each message, or that a message might not be delivered?

Reply:
> 
> 
> What should happen if the endpoint specified does not meet these expectations?
> 
> When might the system discover this?

This situation will be discovered during the event delivery. And this will be reflected in the audit log of the mod-pubsub.

Reply:
> What are the expectations that `mod-pubsub` has for the response from a delivery endpoint?

Any 2xx response is Ok. A response body does not matter for mod-pubsub. Any other response is considered an error.

Reply:
> 
> 
> If delivery fails, will `mod-pubsub` attempt to retry?
> 
> Potentially delivering a message more than once, meaning subscribers might need to be idempotent for each message, or that a message might not be delivered?

In the first version, we will not provide redelivery.

Reply:
Given that the notification for TTL is that no subscribers have received the published event with the time limit and there are no retries.

If there are 3 subscribers, the first receives a copy of the event within the time limit, the other two produce errors.

Does the publisher have any knowledge of those subsequent failures?